go/vt/vttablet/tabletserver/txthrottler/tx_throttler.go

```diff
@@ -387,15 +471,48 @@ func (ts *txThrottlerStateImpl) healthChecksProcessor(ctx context.Context, topoS
 	}
 }
 
-func (ts *txThrottlerStateImpl) throttle() bool {
+func checkEnginePoolUsage(engine TabletserverEngineInterface, thresholds *flagutil.LowHighFloat64Values, highErr, lowErr error) error {
+	// Calls to .GetPoolUsagePercent() are serialized by the underlying engine.
+	switch usagePercent := engine.GetPoolUsagePercent(); {
+	case thresholds.High != 0 && usagePercent >= thresholds.High:
+		return highErr
+	case thresholds.Low != 0 && usagePercent >= thresholds.Low:
+		return lowErr
+	default:
+		return nil
+	}
+}
+
+func (ts *txThrottlerStateImpl) throttle(plan *planbuilder.Plan) error {
 	if ts.throttler == nil {
 		log.Error("txThrottler: throttle called after deallocateResources was called")
-		return false
+		return nil
+	}
+	if plan == nil {
+		log.Error("txThrottler: throttle called with no plan")
+		return nil
+	}
+	switch plan.PlanID {
+	case planbuilder.PlanSelect, planbuilder.PlanSelectImpossible, planbuilder.PlanShow:
+		// check query engine conn pool usage
+		return checkEnginePoolUsage(ts.queryEngine, ts.config.TxThrottlerQueryPoolThresholds,
+			ErrThrottledConnPoolUsageHard, ErrThrottledConnPoolUsageSoft)
+	default:
+		// check tx engine pool usage
+		if err := checkEnginePoolUsage(ts.txEngine, ts.config.TxThrottlerTxPoolThresholds,
+			ErrThrottledTxPoolUsageHard, ErrThrottledTxPoolUsageSoft); err != nil {
+			return err
+		}
+
+		// check max replication lag.
+		// serialize calls to ts.throttle.Throttle().
+		ts.throttleMu.Lock()
```

Comment:
⚠️ I'm not sure I like this mutex. It will effectively serialize all `SELECT/INSERT/UPDATE/DELETE` calls, right? If so, I can't find the justification. It's better to hit the throttler concurrently, multiple times, than to linearize all database access.